### PR TITLE
Remove gnmit/ and merge into gnmi/

### PR DESCRIPTION
**Pull request by wenovus**

### CODE REVIEW

File: gnmi/gnmi.go [hunk 1]
```diff
@@ -20,23 +20,15 @@ import (
...
-	*gnmit.GNMIServer
-	s            *grpc.Server
-	Responses    [][]*gnmipb.SubscribeResponse
-	GetResponses []interface{}
-	Errs         []error
+	*gnmit.GNMIServer[*oc.Root]
 }
 
-func createGNMIServer(targetName string) (*gnmit.GNMIServer, error) {
+func New(srv *grpc.Server, targetName string) (*Server, error) {
```
Comment (DanG100) [inline gnmi/gnmi.go:31]: probably just delete this and move it lemming

**Comment by wenovus** [comment]
> i think i like having everything moved to gnmi and getting rid of gnmit.

I don't like having everything in the `gnmi` package since the name seems too generic to me (are we saying this is "the" gNMI implementation?)

Also right now `gnmit` is generic to any schema, whereas `gnmi` is specific to lemming's GoStructs, so I'm not sure if they should be in the same namespace. What do you think?

File: gnmi/gnmi.go (hunk 1 above)
Comment (wenovus) [reply gnmi/gnmi.go:31]: I kept this to keep uniformity with the other protocols

File: gnmi/gnmit/gnmit.go [hunk 2]
```diff
@@ -157,6 +167,10 @@ func SetupSchema(schema *ytypes.Schema) error {
 		return fmt.Errorf("default root of input schema fails validation: %v", err)
 	}
 
+	if _, ok := schema.Root.(T); !ok {
+		return fmt.Errorf("schema type %T doesn't match generic type %T", schema.Root, new(T))
```
Comment (DanG100) [inline gnmi/gnmit/gnmit.go:171]: ideally this should a compile time check

File: gnmi/gnmit/gnmit.go [hunk 3]
```diff
@@ -58,10 +58,14 @@ const (
 )
 
 // GNMIServer implements the gNMI server interface.
-type GNMIServer struct {
+type GNMIServer[T ygot.GoStruct] struct {
```
Comment (DanG100) [inline gnmi/gnmit/gnmit.go:61]: do we really need to support different schemas? could this just use the oc.Schema

**Comment by DanG100** [comment]
> > i think i like having everything moved to gnmi and getting rid of gnmit.
> 
> I don't like having everything in the `gnmi` package since the name seems too generic to me (are we saying this is "the" gNMI implementation?)
> 
> Also right now `gnmit` is generic to any schema, whereas `gnmi` is specific to lemming's GoStructs, so I'm not sure if they should be in the same namespace. What do you think?

How about

gnmi.New() Server[*oc.Schema]
gnmi.NewWithSchema[T ygot.GoStruct]....

or do even need to support different schemas...

File: gnmi/gnmi.go [hunk 4]
```diff
@@ -17,26 +17,65 @@ package gnmi
...
+
+	configMu     sync.Mutex
+	configSchema *ytypes.Schema
+
+	stateMu     sync.Mutex
+	stateSchema *ytypes.Schema
 }
 
-func createGNMIServer(targetName string) (*gnmit.GNMIServer, error) {
+// New creates and registers a fake gNMI server on the given gRPC server.
```
[the remark this answers is not in the document]
Comment (wenovus) [reply gnmi/gnmi.go:75]: How about I just say "reference gNMI server" (it was called fake in the initial version of the file).

File: gnmi/gnmi.go [hunk 5]
```diff
@@ -45,33 +84,262 @@ func createGNMIServer(targetName string) (*gnmit.GNMIServer, error) {
...
-func New(s *grpc.Server, targetName string) (*Server, error) {
-	gnmiServer, err := createGNMIServer(targetName)
+// NewServer returns a new fake gNMI server implementation that can be registered on
+// an existing gRPC server.
+//
+// - configSchema is the specification of the schema if gnmi.Set on config paths is used.
+// - stateSchema is the specification of the schema if gnmi.Set on state paths is used.
+// - targetName is the name of the target.
+// - sendMeta indicates whether metadata should be sent
+func NewServer(ctx context.Context, configSchema *ytypes.Schema, stateSchema *ytypes.Schema, targetName string, sendMeta bool) (*Collector, *Server, error) {
```
[the remark this answers is not in the document]
Comment (wenovus) [reply gnmi/gnmi.go]: unexported and removed schema inputs.

File: gnmi/gnmi.go [hunk 6]
```diff
@@ -45,33 +84,262 @@ func createGNMIServer(targetName string) (*gnmit.GNMIServer, error) {
...
+	s.s = srv
+	gpb.RegisterGNMIServer(srv, s)
+}
+
+// Start starts the collector-backed gNMI server that listens on the specified
+// addr (in the form host:port).
+//
+// It returns the address it is listening on in the form hostname:port or any
+// errors encounted whilst setting it up.
+func (s *Server) Start(ctx context.Context, addr string, opts ...grpc.ServerOption) (string, error) {
```
[the remark this answers is not in the document]
Comment (wenovus) [reply gnmi/gnmi.go]: moved to test helper

File: gnmi/gnmi.go [hunk 7]
```diff
@@ -45,33 +84,262 @@ func createGNMIServer(targetName string) (*gnmit.GNMIServer, error) {
...
+		configSchema: configSchema,
+		stateSchema:  stateSchema,
+	}
+
+	return c, gnmiserver, nil
+}
+
+// RegisterGRPCServer registers the Server on the provided gRPC Server.
+func (s *Server) RegisterGRPCServer(srv *grpc.Server) {
+	s.s = srv
```
[the remark this answers is not in the document]
Comment (wenovus) [reply gnmi/gnmi.go]: Removed.

**Review by DanG100 — APPROVED** [review]

**Pull request merged**